src/Gaffer/Context.cpp

```diff
@@ -574,31 +589,22 @@ const std::string &Context::SubstitutionProvider::variable( const boost::string_
 	return m_formattedString;
 }
 
-void Context::validateHashes()
+void Context::validateVariableHash( const Storage &s, const IECore::InternedString &name ) const
 {
-	bool origHashValid = m_hashValid;
-	IECore::MurmurHash origHash = m_hash;
+	IECore::MurmurHash prevHash = s.hash;
 
-	// TODO - this is now unsafe, because it involves reading the entries during destruction,
-	// and we probably don't want to require the caller to destruct the EditScope before anything
-	// it might reference.  Should probably switch to validating during get() instead
+	TypeFunctionTable::updateHash( s.typeId, const_cast<Storage&>( s ), name );
 
-	for( Map::iterator it = m_map.begin(), eIt = m_map.end(); it != eIt; ++it )
+	if( prevHash != s.hash )
 	{
-		IECore::MurmurHash prevHash = it->second.hash;
-
-		// TODO - even for something used only for debugging, maybe this should be less inefficient
-		ConstDataPtr asData = TypeFunctionTable::makeData( it->second.typeId, it->second.data );
-		TypeFunctionTable::internalSet( it->second.typeId, *this, it->first, asData.get() );
-
-		if( prevHash != it->second.hash )
-		{
-			throw IECore::Exception( "Corrupt hash for context entry: " + it->first.string() );
-		}
+		throw IECore::Exception( "Corrupt hash for context entry: " + name.string() );
 	}
-
-	if( origHashValid )
+}
+void Context::validateHashes()
+{
+	if( m_hashValid )
 	{
+		IECore::MurmurHash origHash = m_hash;
 		if( origHash != hash() )
```

Comment:
Good point that this isn't catching anything.  In cfd831c I've updated this to set m_hashValid to false before calling hash(), so it does check something ... but it's effectively only checking that nothing modifies a stored hash without setting m_hashValid to false.  This would be an internal error with private data that someone can't trigger from outside, so we should probably go ahead and remove this check entirely.

Reply:
Makes sense - let's remove then.